akka-http-core/src/main/scala/akka/http/impl/engine/ws/Masking.scala

```diff
@@ -50,42 +53,64 @@ private[http] object Masking {
   }
 
   /** Implements both masking and unmasking which is mostly symmetric (because of XOR) */
-  private abstract class Masker extends StatefulStage[FrameEvent, FrameEventOrError] {
+  private abstract class Masker extends GraphStage[FlowShape[FrameEvent, FrameEventOrError]] {
     def extractMask(header: FrameHeader): Int
     def setNewMask(header: FrameHeader, mask: Int): FrameHeader
 
-    def initial: State = Idle
-
-    private object Idle extends State {
-      def onPush(part: FrameEvent, ctx: Context[FrameEventOrError]): SyncDirective =
-        part match {
-          case start @ FrameStart(header, data) ⇒
-            try {
-              val mask = extractMask(header)
-              become(new Running(mask))
-              current.onPush(start.copy(header = setNewMask(header, mask)), ctx)
-            } catch {
-              case p: ProtocolException ⇒
-                become(Done)
-                ctx.push(FrameError(p))
-            }
-          case _: FrameData ⇒
-            ctx.fail(new IllegalStateException("unexpected FrameData (need FrameStart first)"))
+    private val in = Inlet[FrameEvent](s"${toString}-in")
+    private val out = Outlet[FrameEventOrError](s"${toString}-out")
+    override val shape: FlowShape[FrameEvent, FrameEventOrError] = FlowShape(in, out)
+
+    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
+
+      val initialHandler = new InHandler {
```

Comment:
Is only used once but kept alive during the lifetime of the stage by being a class field, make it a def!